src/libraries/System.Net.WebSockets/tests/ZLibStreamTests.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Net.WebSockets.Tests
+{
+    public class ZLibStreamTests
+    {
+        [Fact]
+        public async Task PoolShouldReuseTheSameInstance()
+        {
+            var pool = new Pool(timeoutMilliseconds: 100);
+
+            object inflater = pool.GetInflater();
+            for ( var i = 0; i < 10_000; ++i)
+            {
+                pool.ReturnInflater(inflater);
+                Assert.Equal(1, pool.ActiveCount);
+
+                object nextInflater = pool.GetInflater();
+                Assert.Equal(0, pool.ActiveCount);
+
+                Assert.Equal(inflater, nextInflater);
+            }
+            pool.ReturnInflater(inflater);
+
+            Assert.Equal(1, pool.ActiveCount);
+            await Task.Delay(250);
+
+            // After timeout elapses we should not have any active instances
+            Assert.Equal(0, pool.ActiveCount);
+        }
+
+        [Fact]
+        [PlatformSpecific(~TestPlatforms.Browser)] // There is no concurrency in browser
+        public async Task PoolingConcurrently()
+        {
+            var pool = new Pool(timeoutMilliseconds: 100);
+            var parallelOptions = new ParallelOptions
+            {
+                MaxDegreeOfParallelism = 16
+            };
+            Parallel.For(0, 100_000, parallelOptions, x =>
+            {
+                if (x % 2 == 0)
+                {
+                    object inflater = pool.GetInflater();
+                    pool.ReturnInflater(inflater);
+                }
+                else
+                {
+                    object deflater = pool.GetDeflater();
+                    pool.ReturnDeflater(deflater);
+                }
+            });
+
+            Assert.True(pool.ActiveCount >= 2);
+            Assert.True(pool.ActiveCount <= parallelOptions.MaxDegreeOfParallelism * 2);
+            await Task.Delay(250);
+            Assert.Equal(0, pool.ActiveCount);
+        }
+
+        private sealed class Pool
```

Comment:
We don't use InternalsVisibleTo. Is there some other way we can test?

Reply:
This is why I didn't use InternalsVisibleTo - I remember reading somewhere that InternalsVisibleTo interfere with the linker's ability to trim down assemblies? I could go down the route and include the file as shared source, but this is not enough - I will need to include the other zlib files as well and will need to repurpose the csproj for cross platform build because of native deps. I think the reflection isn't too bad comparing to this.

Reply:
I've seen the usages of InternalsVisibleTo in some places like Microsoft.Extensions.Http and System.IO.Pipelines, I guess they don't care about it for some reason... whatever. If there's no better way to test, then it's ok. I agree that the shared code route might be worse.